File: src/main/java/com/humanizar/units/model/enums/ReasonCode.java
package com.humanizar.units.model.enums;

public enum ReasonCode {

    UNIT_NOT_FOUND(404, "Unidade não encontrada.", false),
    MUNICIPIO_NOT_FOUND(404, "Município não encontrado.", false),
    MUNICIPIO_DUPLICATED(409, "Município já cadastrado.", false),
    MUNICIPIO_HAS_UNITS(409, "Município possui unidades vinculadas.", false),
    VALIDATION_ERROR(400, "Falha de validação da solicitação.", false),
    AUTHENTICATION_FAILURE(401, "Falha de autenticação: credenciais ausentes ou inválidas.", false),
    AUTHORIZATION_FAILURE(403, "Falha de autorização: acesso negado para este recurso.", false),
    PERSISTENCE_FAILURE(503, "Falha de persistência no banco de dados.", true);

    private final int statusCode;
    private final String message;
    private final boolean retryable;

    ReasonCode(int statusCode, String message, boolean retryable) {
        this.statusCode = statusCode;
        this.message = message;
        this.retryable = retryable;
    }

    public int getStatusCode() {
        return statusCode;
    }

    public String getMessage() {
        return message;
    }

    public boolean isRetryable() {
        return retryable;
    }
}

File: src/domain/model/enums/reason_code.rs
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum ReasonCode {
    InvalidRequest,
    ValidationError,
    AuthenticationFailure,
    AuthorizationFailure,
    UnitNotFound,
    MunicipioNotFound,
    UnitDuplicated,
    MunicipioDuplicated,
    MunicipioHasUnits,
    UnexpectedError,
    PersistenceFailure,
}

impl ReasonCode {
    pub const fn code(self) -> &'static str {
        match self {
            Self::InvalidRequest => "INVALID_REQUEST",
            Self::ValidationError => "VALIDATION_ERROR",
            Self::AuthenticationFailure => "AUTHENTICATION_FAILURE",
            Self::AuthorizationFailure => "AUTHORIZATION_FAILURE",
            Self::UnitNotFound => "UNIT_NOT_FOUND",
            Self::MunicipioNotFound => "MUNICIPIO_NOT_FOUND",
            Self::UnitDuplicated => "UNIT_DUPLICATED",
            Self::MunicipioDuplicated => "MUNICIPIO_DUPLICATED",
            Self::MunicipioHasUnits => "MUNICIPIO_HAS_UNITS",
            Self::UnexpectedError => "UNEXPECTED_ERROR",
            Self::PersistenceFailure => "PERSISTENCE_FAILURE",
        }
    }

    pub const fn message(self) -> &'static str {
        match self {
            Self::InvalidRequest => "Requisição inválida.",
            Self::ValidationError => "Falha de validação da solicitação.",
            Self::AuthenticationFailure => {
                "Falha de autenticação: credenciais ausentes ou inválidas."
            }
            Self::AuthorizationFailure => "Falha de autorização: acesso negado para este recurso.",
            Self::UnitNotFound => "Unidade não encontrada.",
            Self::MunicipioNotFound => "Município não encontrado.",
            Self::UnitDuplicated => "Unidade com este CNPJ já cadastrada no município.",
            Self::MunicipioDuplicated => "Município já cadastrado.",
            Self::MunicipioHasUnits => "Município possui unidades vinculadas.",
            Self::UnexpectedError => "Erro inesperado.",
            Self::PersistenceFailure => "Falha de persistência no banco de dados.",
        }
    }

    pub const fn is_retryable(self) -> bool {
        matches!(self, Self::PersistenceFailure)
    }
}
